File: packages/prisma-accumulator/src/operations/read.ts
import type {
  Register,
  BalanceOptions,
  TurnoverOptions,
  MovementsOptions,
  ResourceResult,
  TurnoverRow,
  MovementRow,
} from '../types/register';
import type { DimensionFilter } from '../types/pg-types';
import { toTimestamp, sqlIdentifier } from '../utils';
import { mapPgError } from '../errors';

type Executor = { $queryRawUnsafe: (...args: unknown[]) => Promise<unknown[]> };

/**
 * Query current or historical balance.
 */
export async function balance<
  D extends Record<string, string>,
  R extends Record<string, string>,
>(
  executor: Executor,
  schema: string,
  register: Register<D, R>,
  dims?: DimensionFilter<D>,
  options?: BalanceOptions,
): Promise<ResourceResult<R> | null> {
  const name = register._def.name;
  const fnName = `${name}_balance`;

  const params: unknown[] = [];
  const parts: string[] = [];
  let idx = 1;

  if (dims && Object.keys(dims).length > 0) {
    parts.push(`dimensions := $${idx}::jsonb`);
    params.push(JSON.stringify(dims));
    idx++;
  }

  if (options?.atDate) {
    parts.push(`at_date := $${idx}::timestamptz`);
    params.push(toTimestamp(options.atDate));
    idx++;
  }

  const argList = parts.length > 0 ? parts.join(', ') : '';
  const sql = `SELECT * FROM "${schema}".${sqlIdentifier(fnName)}(${argList})`;

  try {
    const rows = await executor.$queryRawUnsafe(sql, ...params);
    if (!rows || rows.length === 0) return null;
    const row = rows[0] as Record<string, unknown>;
    // The function returns a single JSONB column or flattened resources
    const result = row[fnName] ?? row;
    if (typeof result === 'string') {
      return JSON.parse(result) as ResourceResult<R>;
    }
    return result as ResourceResult<R>;
  } catch (err) {
    mapPgError(err);
  }
}

/**
 * Query turnover for a period.
 */
export async function turnover<
  D extends Record<string, string>,
  R extends Record<string, string>,
>(
  executor: Executor,
  schema: string,
  register: Register<D, R>,
  dims?: DimensionFilter<D>,
  options?: TurnoverOptions,
): Promise<TurnoverRow<R>[]> {
  const name = register._def.name;
  const fnName = `${name}_turnover`;

  const params: unknown[] = [];
  const parts: string[] = [];
  let idx = 1;

  if (options?.dateFrom) {
    parts.push(`from_date := $${idx}::timestamptz`);
    params.push(toTimestamp(options.dateFrom));
    idx++;
  }

  if (options?.dateTo) {
    parts.push(`to_date := $${idx}::timestamptz`);
    params.push(toTimestamp(options.dateTo));
    idx++;
  }

  if (dims && Object.keys(dims).length > 0) {
    parts.push(`dimensions := $${idx}::jsonb`);
    params.push(JSON.stringify(dims));
    idx++;
  }

  if (options?.groupBy && options.groupBy.length > 0) {
    parts.push(`group_by := $${idx}::jsonb`);
    params.push(JSON.stringify(options.groupBy));
    idx++;
  }

  const argList = parts.length > 0 ? parts.join(', ') : '';
  const sql = `SELECT * FROM "${schema}".${sqlIdentifier(fnName)}(${argList})`;

  try {
    const rows = await executor.$queryRawUnsafe(sql, ...params);
    return (rows as Record<string, unknown>[]).map((row) => {
      const val = row[fnName] ?? row;
      if (typeof val === 'string') return JSON.parse(val);
      return val;
    }) as TurnoverRow<R>[];
  } catch (err) {
    mapPgError(err);
  }
}

/**
 * Query movements with filters and pagination.
 */
export async function movements<
  D extends Record<string, string>,
  R extends Record<string, string>,
>(
  executor: Executor,
  schema: string,
  register: Register<D, R>,
  dims?: DimensionFilter<D>,
  options?: MovementsOptions,
): Promise<MovementRow<D, R>[]> {
  const name = register._def.name;
  const fnName = `${name}_movements`;

  const params: unknown[] = [];
  const parts: string[] = [];
  let idx = 1;

  if (options?.recorder) {
    parts.push(`p_recorder := $${idx}`);
    params.push(options.recorder);
    idx++;
  }

  if (options?.dateFrom) {
    parts.push(`from_date := $${idx}::timestamptz`);
    params.push(toTimestamp(options.dateFrom));
    idx++;
  }

  if (options?.dateTo) {
    parts.push(`to_date := $${idx}::timestamptz`);
    params.push(toTimestamp(options.dateTo));
    idx++;
  }

  if (dims && Object.keys(dims).length > 0) {
    parts.push(`dimensions := $${idx}::jsonb`);
    params.push(JSON.stringify(dims));
    idx++;
  }

  const argList = parts.length > 0 ? parts.join(', ') : '';
  let sql = `SELECT * FROM "${schema}".${sqlIdentifier(fnName)}(${argList})`;

  if (options?.limit) {
    sql += ` LIMIT ${Number(options.limit)}`;
  }
  if (options?.offset) {
    sql += ` OFFSET ${Number(options.offset)}`;
  }

  try {
    const rows = await executor.$queryRawUnsafe(sql, ...params);
    return (rows as Record<string, unknown>[]).map((row) => {
      const val = row[fnName] ?? row;
      if (typeof val === 'string') return JSON.parse(val);
      return val;
    }) as MovementRow<D, R>[];
  } catch (err) {
    mapPgError(err);
  }
}

File: packages/sqlalchemy-accumulator/src/sqlalchemy_accumulator/operations/read.py
"""Read operations: balance, turnover, movements."""

from __future__ import annotations

import json
from decimal import Decimal
from typing import Any

from sqlalchemy import text
from sqlalchemy.engine import Connection

from ..errors import map_pg_error
from ..types import Register, BalanceOptions, TurnoverOptions, MovementsOptions
from ..validation import sql_identifier, to_timestamp


def balance(
    conn: Connection,
    schema: str,
    register: Register,
    dims: dict[str, Any] | None = None,
    options: BalanceOptions | None = None,
) -> dict[str, Decimal] | None:
    """Query current or historical balance.

    Returns a dict of ``{resource_name: value}`` or ``None`` if no data found.
    """
    name = register._def.name
    fn_name = f"{name}_balance"

    params: dict[str, Any] = {}
    parts: list[str] = []

    if dims:
        parts.append("dimensions := :dims::jsonb")
        params["dims"] = json.dumps(dims, default=str)

    if options and options.at_date is not None:
        parts.append("at_date := :at_date::timestamptz")
        params["at_date"] = to_timestamp(options.at_date)

    arg_list = ", ".join(parts)
    sql = f'SELECT * FROM "{schema}".{sql_identifier(fn_name)}({arg_list})'

    try:
        result = conn.execute(text(sql), params)
        row = result.mappings().fetchone()
        if not row:
            return None

        # The PG function returns a single JSONB column named after the function,
        # or flattened resource columns. Handle both cases.
        if fn_name in row:
            val = row[fn_name]
            if isinstance(val, str):
                parsed = json.loads(val)
                return {k: Decimal(str(v)) for k, v in parsed.items()}
            if isinstance(val, dict):
                return {k: Decimal(str(v)) for k, v in val.items()}
            return None

        # Flattened columns — extract resource keys
        return {
            k: Decimal(str(row[k]))
            for k in register._def.resources
            if k in row
        }
    except Exception as exc:
        map_pg_error(exc)
        raise


def turnover(
    conn: Connection,
    schema: str,
    register: Register,
    dims: dict[str, Any] | None = None,
    options: TurnoverOptions | None = None,
) -> list[dict[str, Any]]:
    """Query turnover for a period. Returns list of result dicts."""
    name = register._def.name
    fn_name = f"{name}_turnover"

    params: dict[str, Any] = {}
    parts: list[str] = []

    if options and options.date_from is not None:
        parts.append("from_date := :date_from::timestamptz")
        params["date_from"] = to_timestamp(options.date_from)

    if options and options.date_to is not None:
        parts.append("to_date := :date_to::timestamptz")
        params["date_to"] = to_timestamp(options.date_to)

    if dims:
        parts.append("dimensions := :dims::jsonb")
        params["dims"] = json.dumps(dims, default=str)

    if options and options.group_by:
        parts.append("group_by := :group_by::jsonb")
        params["group_by"] = json.dumps(options.group_by)

    arg_list = ", ".join(parts)
    sql = f'SELECT * FROM "{schema}".{sql_identifier(fn_name)}({arg_list})'

    try:
        result = conn.execute(text(sql), params)
        rows = result.mappings().fetchall()
        out: list[dict[str, Any]] = []
        for row in rows:
            if fn_name in row:
                val = row[fn_name]
                if isinstance(val, str):
                    out.append(json.loads(val))
                elif isinstance(val, dict):
                    out.append(val)
                else:
                    out.append(dict(row))
            else:
                out.append(dict(row))
        return out
    except Exception as exc:
        map_pg_error(exc)
        raise


def movements(
    conn: Connection,
    schema: str,
    register: Register,
    dims: dict[str, Any] | None = None,
    options: MovementsOptions | None = None,
) -> list[dict[str, Any]]:
    """Query movements with filters and pagination."""
    name = register._def.name
    fn_name = f"{name}_movements"

    params: dict[str, Any] = {}
    parts: list[str] = []

    if options and options.recorder is not None:
        parts.append("p_recorder := :recorder")
        params["recorder"] = options.recorder

    if options and options.date_from is not None:
        parts.append("from_date := :date_from::timestamptz")
        params["date_from"] = to_timestamp(options.date_from)

    if options and options.date_to is not None:
        parts.append("to_date := :date_to::timestamptz")
        params["date_to"] = to_timestamp(options.date_to)

    if dims:
        parts.append("dimensions := :dims::jsonb")
        params["dims"] = json.dumps(dims, default=str)

    arg_list = ", ".join(parts)
    sql = f'SELECT * FROM "{schema}".{sql_identifier(fn_name)}({arg_list})'

    if options and options.limit is not None:
        sql += f" LIMIT {int(options.limit)}"
    if options and options.offset is not None:
        sql += f" OFFSET {int(options.offset)}"

    try:
        result = conn.execute(text(sql), params)
        rows = result.mappings().fetchall()
        out: list[dict[str, Any]] = []
        for row in rows:
            if fn_name in row:
                val = row[fn_name]
                if isinstance(val, str):
                    out.append(json.loads(val))
                elif isinstance(val, dict):
                    out.append(val)
                else:
                    out.append(dict(row))
            else:
                out.append(dict(row))
        return out
    except Exception as exc:
        map_pg_error(exc)
        raise
